pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -0,0 +1,743 @@
+	}
+
+	return nil
+}
+
+// PostFilter checks whether freeing an allocated claim might help to get a Pod
+// schedulable. This only gets called when filtering found no suitable node.
+func (pl *dynamicResources) PostFilter(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, filteredNodeStatusMap framework.NodeToStatusMap) (*framework.PostFilterResult, *framework.Status) {
+	logger := klog.FromContext(ctx)
+	state, err := getStateData(cs)
+	if err != nil {
+		return nil, statusError(logger, err)
+	}
+	if len(state.claims) == 0 {
+		return nil, nil
+	}
+
+	// Iterating over a map is random. This is intentional here, we want to
+	// pick one claim randomly because there is no better heuristic.
+	for index := range state.unavailableClaims {
+		claim := state.claims[index]
+		if len(claim.Status.ReservedFor) == 0 ||
+			len(claim.Status.ReservedFor) == 1 && claim.Status.ReservedFor[0].UID == pod.UID {
+			claim := state.claims[index].DeepCopy()
+			claim.Status.DeallocationRequested = true
+			claim.Status.ReservedFor = nil
+			logger.V(5).Info("reallocate", "pod", klog.KObj(pod), "resourceclaim", klog.KObj(claim))
+			if err := state.updateClaimStatus(ctx, index, claim); err != nil {
+				return nil, statusError(logger, err)
+			}
+			break
+		}
+	}
+	return nil, nil
+}
+
+// PreScore is passed a list of all nodes that would fit the pod. Not all
+// claims are necessarily allocated yet, so here we can set the SuitableNodes
+// field for those which are pending.
+func (pl *dynamicResources) PreScore(ctx context.Context, cs *framework.CycleState, pod *v1.Pod, nodes []*v1.Node) *framework.Status {
+	state, err := getStateData(cs)
+	if err != nil {
+		return statusError(klog.FromContext(ctx), err)
+	}
+	if len(state.claims) == 0 {
+		return nil
+	}
+
+	logger := klog.FromContext(ctx)
+	podScheduling, err := state.initializePodScheduling(ctx, pod, pl.podSchedulingLister)
+	if err != nil {
+		return statusError(logger, err)
+	}
+	pending := false
+	for _, claim := range state.claims {
+		if claim.Status.Allocation == nil {
+			pending = true
+		}
+	}
+	if pending && !haveAllNodes(podScheduling.Spec.PotentialNodes, nodes) {
+		// Remember the potential nodes. The object will get created or
+		// updated in Reserve. This is both an optimization and
+		// covers the case that PreScore doesn't get called when there
+		// is only a single node.
+		logger.V(5).Info("remembering potential nodes", "pod", klog.KObj(pod), "potentialnodes", klog.KObjSlice(nodes))
+		podScheduling = podScheduling.DeepCopy()
+		numNodes := len(nodes)
+		if numNodes > v1.PodSchedulingNodeListMaxSize {
+			numNodes = v1.PodSchedulingNodeListMaxSize
+		}
+		podScheduling.Spec.PotentialNodes = make([]string, 0, numNodes)
+		if numNodes == len(nodes) {
+			// Copy all node names.
+			for _, node := range nodes {
+				podScheduling.Spec.PotentialNodes = append(podScheduling.Spec.PotentialNodes, node.Name)
+			}
+		} else {
+			// Select a random subset of the nodes to comply with
```

Comment:
Multiple thoughts here:
1. does PotentialNodes have to include SelectedNode?
2. In large clusters, the scheduler only runs filter for a subset of nodes. This means that the list of PotentialNodes can change every cycle.

I imagine this scenario:
1. kube-scheduler evaluates nodes 1-100 and adds the potential nodes
2. driver starts evaluating the nodes.
3. kube-scheduler evaluates nodes 101-200 and updates the potential nodes
4. podscheduling is now updated with the information about nodes 1-100, where they don't fit.
5. kube-scheduler evaluates nodes 201-300 and updates the potential nodes
6. podscheduling is now updated with the information about nodes 101-200, where they don't fit.
7. kube-scheduler evaluates nodes 1-100 again, but podscheduling no longer has the information that node of the nodes 1-100 fit.

Infinite loop.

Reply:
PotentialNodes doesn't have to include SelectedNode, although I would expect that it does.

My hope is that by trying multiple times, the scheduler will eventually be lucky. We will have to evaluate this, but please let's proceed for now because it is already useful for a wide range of clusters, okay?

Reply:
That's not a very encouraging answer. It's this kind of contention that makes me very nervous about proceeding with this API and implementation to beta. We need some kind of guarantee that the scheduling decision converges, at least for the case where the cluster doesn't change.

Reply:
We will not proceed to beta without an answer for this. But right now we are talking about alpha. IMHO the advantages of merging as-is (getting it to users, being able to do incremental improvements) outweigh the risk of having to change the PodScheduling API later on.

Reply:
I can imagine alternatives to the current PodScheduling API:
- let the scheduler post PotentialNodes in several different objects, all linked to the pod through some label instead of using the same name
- let the drivers reply also in different objects

This makes it possible to scale up the number of node names while still maintaining a limit per object.

The important aspect here is that this is entirely between scheduler and drivers and won't affect how users define resource requirements. If driver authors use the common utility code to interface with Kubernetes, then they don't need to adapt their code at all because that common utility code and do the job of joining and splitting node name slices.

Reply:
This is not just about the API limit, but how kube-scheduler operates. If multiple nodes already pass the filters, kube-scheduler stops looking at the remaining nodes. We need a communication protocol between kube-scheduler and the drivers that can work under these circumstances.

Reply:
Once we allow the driver to provide (eventually) information about all unsuitable nodes, the scheduler will be left with just the nodes that really can work.

Reply:
I am confused about step 1-3 - why would the scheduler update when the driver(s) have not responded?

1) kube-scheduler evaluates nodes 1-100 and adds the potential nodes
2) driver starts evaluating the nodes.
3) kube-scheduler evaluates nodes 101-200 and updates the potential nodes

Why #3?  Why is scheduler not waiting for driver?

Reply:
I can imagine that this could happen if some other event triggered another pod scheduling cycle. It looks like "bad luck" is needed repeatedly for the node candidates to flip back and forth like this, which is why I think that it will work better in practice.

A litmus test for the approach will be "fill up a large cluster where each node has local resources". This is one of the things that we'll have to try out as we proceed.

Reply:
That is true: assuming the event handlers are correctly filtering, step 3 wouldn't happen unless there is more than one resource claim.

But even if we assume that step 3 happens after the driver has responded. I see these possibilities:
A. driver successfully allocated in SelectedNode. Since the allocation has a node selector, the scheduler would be directed to the correct node.
B. the claim can't be allocated in any of the PotentialNodes or SelectedNode. Also good: the scheduler would look at the next set of nodes.
C. driver can't allocate in SelectedNode, but some nodes in PotentialNodes would have been possible. Bad: the scheduler might decide to evaluate a different set of nodes that pass the filter. This is the scenario that can repeat almost indefinitely.

> A litmus test for the approach will be "fill up a large cluster where each node has local resources".

And multiple pods asking for resources, and 90% of them have the local resources used (not a crazy thing to happen in a highly utilized cluster). Then, measure latency.

Reply:
This would be the main blocker for graduation to beta from sig-scheduling PoV.